src/Microsoft.AspNetCore.Mvc.Razor.Host/Internal/ViewComponentTagHelperChunkVisitor.cs

```diff
@@ -0,0 +1,228 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Razor.Chunks;
+using Microsoft.AspNetCore.Razor.CodeGenerators;
+using Microsoft.AspNetCore.Razor.CodeGenerators.Visitors;
+using Microsoft.AspNetCore.Razor.Compilation.TagHelpers;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Host.Internal
+{
+    /// <summary>
+    /// A <see cref="CodeVisitor{CSharpCodeWriter}"/> that parses <see cref="Chunk"/>s and 
+    /// generates code based on the contents of the chunks that are view component tag helpers.
+    /// </summary>
+    public class ViewComponentTagHelperChunkVisitor : CodeVisitor<CSharpCodeWriter>
+    {
+        private GeneratedViewComponentTagHelperContext _context;
+        private HashSet<string> _writtenViewComponents;
+
+        /// <summary>
+        /// Creates a new instance of <see cref="ViewComponentTagHelperChunkVisitor"/>. 
+        /// </summary>
+        /// <param name="writer">The <see cref="CSharpCodeWriter"/> the code is generated to.</param>
+        /// <param name="context">The <see cref="CodeGeneratorContext"/> needed to visit chunks.</param>
+        public ViewComponentTagHelperChunkVisitor(CSharpCodeWriter writer, CodeGeneratorContext context) :
+            base(writer, context)
+        {
+            if (writer == null)
+            {
+                throw new ArgumentNullException(nameof(writer));
+            }
+
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            _context = new GeneratedViewComponentTagHelperContext();
+            _writtenViewComponents = new HashSet<string>();
+        }
+
+        /// <summary>
+        /// Generates code if the selected <see cref="Chunk"/>, if the chunk is a <see cref="TagHelperChunk"/>
+        /// that describes a view component. If the chunk is a <see cref="ParentChunk"/>, accepts the chunk's children. 
+        /// </summary>
+        /// <param name="chunk">A <see cref="Chunk"/> to accept.</param> 
+        public override void Accept(Chunk chunk)
+        {
+            if (chunk == null)
+            {
+                throw new ArgumentNullException(nameof(chunk));
+            }
+
+            var parentChunk = chunk as ParentChunk;
+            var tagHelperChunk = chunk as TagHelperChunk;
+
+            if (parentChunk != null && !(parentChunk is TagHelperChunk))
+            {
+                Accept(parentChunk.Children);
+            }
+
+            else if (tagHelperChunk != null) 
```

Comment:
If the above if is not true you can just do base.Accept(chunk).

With what you're doing right now you're triggering all of the Visits an additional time per descriptor without doing any extra logic :smile: